File: FSharp/Shared/SftpClient.fs
[<AutoOpen>]
module SftpClient

    open System
    open System.Collections.Generic
    open System.IO

    type SftpFileInfo = {
        Name : string
        Length : int64
        IsDirectory : bool
    }

    type ISftpAsyncResult =
        inherit IAsyncResult
        abstract member AsyncResult : IAsyncResult
        abstract member IsCanceled : bool with get, set

    type ISftpClient =
        abstract member Connect : unit -> unit
        abstract member Disconnect : unit -> unit
        abstract member DownloadFile : string * Stream * Action<uint64> -> unit
        abstract member BeginDownloadFile : string * Stream * AsyncCallback * obj * Action<uint64> -> IAsyncResult
        abstract member EndDownloadFile : IAsyncResult -> unit
        abstract member UploadFile : Stream * string * Action<uint64> -> unit
        abstract member BeginUploadFile : Stream * string * AsyncCallback * obj * Action<uint64> -> IAsyncResult
        abstract member EndUploadFile : IAsyncResult -> unit
        abstract member RenameFile : string * string -> unit
        abstract member DeleteFile : string -> unit
        abstract member CreateDirectory : string -> unit
        abstract member DeleteDirectory : string -> unit
        abstract member ListDirectory : string * Action<int> -> IEnumerable<SftpFileInfo>
        abstract member DirectoryExists : string -> bool
        abstract member Dispose : unit -> unit

    type ISshCommand =
        abstract member Execute : unit -> string
        abstract member BeginExecute : AsyncCallback -> IAsyncResult
        abstract member EndExecute : IAsyncResult -> string
        abstract member CancelAsync : unit -> unit

    type ISshClient =
        abstract member Connect : unit -> unit
        abstract member Disconnect : unit -> unit
        abstract member CreateCommand : string -> ISshCommand
        abstract member Dispose : unit -> unit

    type IFileStreamProvider =
        abstract member OpenRead : string -> Stream
        abstract member OpenWrite : string -> Stream

    type IClientFactory =
        abstract member CreateSftpClient : unit -> ISftpClient
        abstract member CreateSshClient : unit -> ISshClient
        abstract member CreateFileStreamProvider : unit -> IFileStreamProvider
        abstract member CreateSftpAsyncResult : IAsyncResult -> ISftpAsyncResult

    type FileStreamProvider() =
        interface IFileStreamProvider with
            member this.OpenRead(path : string) = File.OpenRead(path) :> Stream
            member this.OpenWrite(path : string) = File.OpenWrite(path) :> Stream

File: CSharp/Shared/SftpClient.cs
using System;
using System.Collections.Generic;
using System.IO;

namespace Shared
{
    public class SftpFileInfo
    {
        public string Name { get; set; }
        public long Length { get; set; }
        public bool IsDirectory { get; set; }
    }

    public interface ISftpAsyncResult : IAsyncResult
    {
        IAsyncResult AsyncResult { get; }
        bool IsCanceled { get; set; }
    }

    public interface ISftpClient
    {
        void Connect();
        void Disconnect();
        void DownloadFile(string path, Stream stream, Action<ulong> progress);
        IAsyncResult BeginDownloadFile(string path, Stream stream, AsyncCallback callback, Action<ulong> progress);
        void EndDownloadFile(IAsyncResult ar);
        void UploadFile(Stream stream, string path, Action<ulong> progress);
        IAsyncResult BeginUploadFile(Stream stream, string path, AsyncCallback callback, Action<ulong> progress);
        void EndUploadFile(IAsyncResult ar);
        void RenameFile(string sourcePath, string destinationPath);
        void DeleteFile(string path);
        void CreateDirectory(string path);
        void DeleteDirectory(string path);
        IEnumerable<SftpFileInfo> ListDirectory(string path, Action<int> progress);
        bool DirectoryExists(string path);
        void Dispose();
    }

    public interface ISshCommand
    {
        string Execute();
        IAsyncResult BeginExecute(AsyncCallback callback);
        string EndExecute(IAsyncResult result);
        void CancelAsync();
    }

    public interface ISshClient
    {
        void Connect();
        void Disconnect();
        ISshCommand CreateCommand(string command);
        void Dispose();
    }

    public interface IFileStreamProvider
    {
        Stream OpenRead(string path);
        Stream OpenWrite(string path);
    }

    public interface IClientFactory
    {
        ISftpClient CreateSftpClient();
        ISshClient CreateSshClient();
        IFileStreamProvider CreateFileStreamProvider();
        ISftpAsyncResult CreateSftpAsyncResult(IAsyncResult result);
    }

    public class FileStreamProvider : IFileStreamProvider
    {
        public Stream OpenRead(string path)
        {
            return File.OpenRead(path);
        }

        public Stream OpenWrite(string path)
        {
            return File.OpenWrite(path);
        }
    }
}
